### PR TITLE
[Fix] Request stack with null request

### LINKED ISSUE #14727: Error : Call to a member function getPathInfo() on null
**Sylius version affected**: >=1.11

**Description**
Using `@serializer` in a cli context.

```
Error : Call to a member function getPathInfo() on null
 vendor/sylius/sylius/src/Sylius/Bundle/ApiBundle/Serializer/HydraErrorNormalizer.php:38
 vendor/symfony/serializer/Serializer.php:278
 vendor/symfony/serializer/Serializer.php:160
 vendor/symfony/serializer/Serializer.php:134
```

```
Error : Call to a member function getPathInfo() on null
 vendor/sylius/sylius/src/Sylius/Bundle/ApiBundle/Serializer/FlattenExceptionNormalizer.php:32
 vendor/symfony/serializer/Serializer.php:278
 vendor/symfony/serializer/Serializer.php:160
 vendor/symfony/serializer/Serializer.php:134
```

**Steps to reproduce**

Create a cli command inject the `@serializer` service and try serializing something :

```php
$this->serializer->serialize($object);
``` 

**Possible Solution**

Fix PHPStan not detected possible null value here : https://github.com/Sylius/Sylius/blob/c1d83248343c78736cd3cfd585901ec6b12fd3f5/src/Sylius/Bundle/ApiBundle/Serializer/HydraErrorNormalizer.php#L37-L42

https://github.com/Sylius/Sylius/blob/c1d83248343c78736cd3cfd585901ec6b12fd3f5/src/Sylius/Bundle/ApiBundle/Serializer/FlattenExceptionNormalizer.php#L31-L36

Also happening with Sylius 1.12 and 1.13 branches.

### PR DESCRIPTION
**Pull request by Prometee**
| Q               | A                                                            |
|-----------------|--------------------------------------------------------------|
| Branch?         | 1.11                  |
| Bug fix?        | yes                                                       |
| New feature?    | no                                                       |
| BC breaks?      | no                                                       |
| Deprecations?   | no |
| Related tickets | fixes #14727                      |
| License         | MIT                                                          |

Add missing PHPSpec tests when `Request` can be `null` when getting it form the `@request_stack` service.

### CODE REVIEW

File: src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php [hunk 1]
```diff
@@ -71,7 +71,9 @@ private function getSession(): SessionInterface
     {
         // bc-layer for Symfony 4
         if (!method_exists(RequestStack::class, 'getSession')) {
-            return $this->requestStack->getMasterRequest()->getSession();
+            $request = $this->requestStack->getMasterRequest();
+            Assert::notNull($request);
+            return $request->getSession();
```
Comment (diimpp) [inline src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php:77]: Missing empty space before `return`

File: src/Sylius/Bundle/AddressingBundle/spec/EventListener/ZoneMemberIntegrityListenerSpec.php [hunk 2]
```diff
@@ -13,12 +13,14 @@
 
 namespace spec\Sylius\Bundle\AddressingBundle\EventListener;
 
+use InvalidArgumentException;
```
Comment (diimpp) [inline src/Sylius/Bundle/AddressingBundle/spec/EventListener/ZoneMemberIntegrityListenerSpec.php]: Implicit PSR rule (via example in PSR, not a actual rule) and usual sylius way is to use it directly in source as `\InvalidArgumentException::class`

File: src/Sylius/Bundle/ApiBundle/Serializer/HydraErrorNormalizer.php [hunk 3]
```diff
@@ -35,13 +36,16 @@ public function normalize($object, $format = null, array $context = [])
...
+            $request = $this->requestStack->getMainRequest();
         } else {
-            /** @phpstan-ignore-next-line */
-            $path = $this->requestStack->getMasterRequest()->getPathInfo();
+            $request = $this->requestStack->getMasterRequest();
         }
 
-        if (!str_starts_with($path, $this->newApiRoute)) {
+        if (null === $request) {
+            return false;
```
Comment (diimpp) [inline src/Sylius/Bundle/ApiBundle/Serializer/HydraErrorNormalizer.php:45]: Shouldn't we assert here, rather than continue as missing master request is an alright situation?

File: src/Sylius/Bundle/ShopBundle/EventListener/OrderCustomerIpListener.php [hunk 4]
```diff
@@ -30,6 +30,13 @@ public function assignCustomerIpToOrder(GenericEvent $event): void
         $subject = $event->getSubject();
         Assert::isInstanceOf($subject, OrderInterface::class);
 
-        $this->ipAssigner->assign($subject, $this->requestStack->getMasterRequest());
+        if (method_exists($this->requestStack, 'getMainRequest')) {
+            $request = $this->requestStack->getMainRequest();
+        } else {
+            $request = $this->requestStack->getMasterRequest();
+        }
+        Assert::notNull($request);
```
Comment (diimpp) [inline src/Sylius/Bundle/ShopBundle/EventListener/OrderCustomerIpListener.php:38]: Please add some error message, webmozart's `notNull` asserts are notorious hard to debug, when stack trace points to vendor/webmozaert/assert/assert and not to offending line of code.

File: src/Sylius/Bundle/AddressingBundle/spec/EventListener/ZoneMemberIntegrityListenerSpec.php (hunk 2 above)
Comment (Prometee) [reply src/Sylius/Bundle/AddressingBundle/spec/EventListener/ZoneMemberIntegrityListenerSpec.php]: I will never understand this rule 😄, for me all classes used in a class file must be in the uses block even if they are part of the root namespace. But if it's a Sylius rule, I will update this 😉

File: src/Sylius/Bundle/ApiBundle/Serializer/HydraErrorNormalizer.php (hunk 3 above)
Comment (Prometee) [reply src/Sylius/Bundle/ApiBundle/Serializer/HydraErrorNormalizer.php:45]: Nope, it's a normalizer and the root cause of this issue #14727

File: src/Sylius/Bundle/ShopBundle/EventListener/OrderCustomerIpListener.php (hunk 4 above)
Comment (Prometee) [reply src/Sylius/Bundle/ShopBundle/EventListener/OrderCustomerIpListener.php:38]: I agree it's always painful 😉

File: src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php (hunk 1 above)
Comment (Prometee) [reply src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php:77]: Is this kind of mistake not covered by the cron task handling ECS ?
https://github.com/Sylius/Sylius/blob/1.13/.github/workflows/refactor.yml

Comment (diimpp) [reply src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php:77]: I'm wondering myself each time I'm seeing it, but ECS check is not perfect right now.

File: src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php [hunk 5]
```diff
@@ -71,7 +71,10 @@ private function getSession(): SessionInterface
     {
         // bc-layer for Symfony 4
         if (!method_exists(RequestStack::class, 'getSession')) {
-            return $this->requestStack->getMasterRequest()->getSession();
+            $request = $this->requestStack->getMasterRequest();
+            Assert::notNull($request, 'No request available to get a Session !');
```
Comment (diimpp) [inline src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php]: > no master(main)* request

I think this is important distinction

**Review by jakubtobiasz — APPROVED** [review]

File: src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php [hunk 6]
```diff
@@ -71,7 +71,10 @@ private function getSession(): SessionInterface
     {
         // bc-layer for Symfony 4
         if (!method_exists(RequestStack::class, 'getSession')) {
-            return $this->requestStack->getMasterRequest()->getSession();
+            $request = $this->requestStack->getMasterRequest();
+            Assert::notNull($request, 'No main request available to get a session !');
```
Comment (coldic3) [inline src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php:75]:
```suggestion
            Assert::notNull($request, 'No main request available to get a session!');
```

**Review by Rafikooo — APPROVED** [review]

**Review by coldic3 — APPROVED** [review]

**Review by NoResponseMate — APPROVED** [review]

**Review by GSadee — APPROVED** [review]

**Comment by GSadee** [comment]
Thanks, Francis! :1st_place_medal:

File: src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php (hunk 6 above)
Comment (Prometee) [reply src/Sylius/Bundle/AddressingBundle/EventListener/ZoneMemberIntegrityListener.php:75]: Ah yes my french syntax is always following me :D
